server/engine/src/bad_frames_detector/thresholds.py

```diff
@@ -355,39 +369,64 @@ def estimate_missing_grfs(self, subject: nimble.biomechanics.SubjectOnDisk, tria
+            if len(trial_proto.getPasses()) < 2 or trial_proto.getPasses()[1].getType() != nimble.biomechanics.ProcessingPassType.ACC_MINIMIZING_FILTER:
+                result.append([nimble.biomechanics.MissingGRFReason.hasInputOutliers] * trial_len)
+                continue
             poses = trial_proto.getPasses()[1].getPoses()
             vels = trial_proto.getPasses()[1].getVels()
 
-            # 3. Check if the trial has badly wrapped IK based on the smoothed velocities
+            # 3. Check if the trial has badly wrapped IK based on the smoothed velocities. In theory, the previous code
+            # should prevent this from happening, but it seems in our dataset that sometimes the IK can still produce
+            # angular wrapping situations, where a joint angle jumps from 0 to 2PI, for example. When we smooth the
+            # position jump, this manifests as unrealistically high velocities.
             if np.max(np.abs(vels)) > 40.0:
                 result.append([nimble.biomechanics.MissingGRFReason.velocitiesStillTooHighAfterFiltering] * trial_len)
                 continue
 
-            # 4. Check for outlying CoP values
+            # 4. We can now begin the more computationally expensive checks. Here, we're checking that the center of
+            # pressure is generally beneath the convex hull outline of a foot. The convex hull is defined by a set of
+            # markers given at the top of the file. If the CoP is outside the convex hull, we mark the frame as bad,
+            # because something is probably wrong with the force plate data. Often this can be a force plate that's
+            # miscalibrated in space, or someone has a bug in their CoP calculation code.
             if self.get_force_weighted_convex_foot_cop_error(skel, foot_markers, poses, raw_force_plate_forces, raw_force_plate_cops) > 0.01:
                 result.append([nimble.biomechanics.MissingGRFReason.copOutsideConvexFootError] * trial_len)
                 continue
 
-            # 5. Estimate the trial type
+            # 5. Estimate the trial type -- this is most important for identifying treadmill trials, which will tend to
+            # have almost all steps on the force plates. Overground trials need further attention.
             trial_type = self.estimate_trial_type(skel, foot_bodies, poses, vels, raw_force_plate_forces, raw_force_plate_cops)
 
             # 6. Check for missing GRFs on footsteps off force plates, for data that is overground and has passed all
-            # the other checks -- For now we just check if the total force magnitude is less than 10 N.
+            # the other checks -- For now we just check if the total force magnitude is less than 10 N. This has the
+            # obvious problem that if you have overground sprinting, we will exclude flight phase frames. Because this
+            # case is so rare, I'm comfortable just asking users to manually annotate those frames, if they care about
```

Comment:
It seems that going forward users will need to interact with the frame annotator in the frontend. It exists there, but we've not done much to explain it to users. Just making a note of it, no changes needed here.

Reply:
Agreed, I think that'll be helpful to explain to them. May need another webinar, and some UI improvements.